Add Highlight In SQL

forestmvey opened the pull request:
### Description
Support highlighting function in SQL query engine.
 
### Issues Resolved
Github Issue: [636](https://github.com/opensearch-project/sql/issues/636)
 
### Check List
- [x] New functionality includes testing.
  - [x] All tests pass, including unit test, integration test and doctest
- [x] New functionality has been documented.
  - [x] New functionality has javadoc added
  - [ ] New functionality has user manual doc added
- [x] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

dai-chen commented on opensearch/src/main/java/org/opensearch/sql/opensearch/storage/OpenSearchIndexScan.java:
```diff
@@ -158,6 +160,22 @@ public void pushDownLimit(Integer limit, Integer offset) {
+    SearchSourceBuilder sourceBuilder = request.getSourceBuilder();
+    if (sourceBuilder.highlighter() != null) {
+      sourceBuilder.highlighter().field(StringUtils.unquoteText(field));
+      sourceBuilder.highlighter(sourceBuilder.highlighter());
```
is this unnecessary?

forestmvey replied:
good catch, i believe this can be removed.

dai-chen commented on core/src/main/java/org/opensearch/sql/expression/HighlightExpression.java:
```diff
@@ -0,0 +1,78 @@
+   * @return : String or Map value of highlight fields.
+   */
+  @Override
+  public ExprValue valueOf(Environment<Expression, ExprValue> valueEnv) {
```
As I understand, `OpenSearch` parses `SearchHits` and add `_highlight.xxx` to value env. Here we try to find `_highlight.xxx` and convert it to `highlight(xxx)` which is specified in `SELECT` by user. I'm thinking is it possible to spare the intermediate format (`_highlight.xxx`)? If so, is this new expression still required?

```
{
   "hits": [{
    "highlight": {
      "name": "..."
      "address": "..."
    }
  }]
}

=>

"_highlight.name"=...
"_highlight.address"=...

=>

highlight("name")
highlight("address")
```

forestmvey replied:
The intermediate format of `_highlight.xxx` is not overly important but we did need some sort of way to parse the highlight fields out of `valueEnv`. In [OpenSearchResponse](https://github.com/opensearch-project/sql/blob/6b96ac04e559160611cd9269fbf08dae02dd372a/opensearch/src/main/java/org/opensearch/sql/opensearch/response/OpenSearchResponse.java#L97-L111) revisions we must create a map of the highlight fields to use the expression and return the list of fields. I don't see a way that we could eliminate the expression entirely and still return single or multiple highlight fields.

dai-chen replied:
Okay. I was thinking in OpenSearchResponse can we return `highlight(field)` directly. Ignore this if impossible. Thanks.

acarbonetto replied:
It just messes up if you pass `highlight("*")`.  We needed a workaround for that.

dai-chen reviewed (commented):
I'm wondering are we sending another PR for user manual separately?

forestmvey commented:
> I'm wondering are we sending another PR for user manual separately?

I added highlight to functions.rst under commit: [232b19f5924d2c46fd08b9d60abadb0d207ad016](https://github.com/opensearch-project/sql/pull/717/commits/232b19f5924d2c46fd08b9d60abadb0d207ad016)

forestmvey commented:
Under commit [9fbea3072bbf6cd3f89851a20344eecf720ac219](https://github.com/opensearch-project/sql/pull/717/commits/9fbea3072bbf6cd3f89851a20344eecf720ac219) I have changed the output format to be more useable to the end user. See revisions to `functions.rst` for new output format. I have also changed highlight all, and wildcard to unsupported for this version as we want to iterate on the output format for multiple returned highlight fields.

acarbonetto reviewed (dismissed):
LGTM

MaxKsyunz commented on core/src/main/java/org/opensearch/sql/expression/HighlightExpression.java:
```diff
@@ -0,0 +1,57 @@
+  }
+
+  /**
+   * Return String or Map value matching highlight field.
```
Is this comment still correct? It can either return a `String` or a `Map`? It seems to always be a collection.

MaxKsyunz commented on core/src/main/java/org/opensearch/sql/expression/HighlightExpression.java:
```diff
@@ -0,0 +1,57 @@
+   * @return : String type.
+   */
+  @Override
+  public ExprType type() {
```
@forestmvey, can you change the return value here to match `ExprValue.type()` returned by `HighlightExpression::valueOf`.

`JDBCFormatter` uses `HighlightExpression::type` to construct the schema object in the response JSON. At the moment it's always `STRING` for highlight fragments.

With `highlight(*)` in mind it may be better to always return `ExprTupleValue`. That way if the user changes their statement from highlighting one field to highlighting many, they will not need to change the code that examines the response.

anirudha commented:
@dai-chen could you review this please
